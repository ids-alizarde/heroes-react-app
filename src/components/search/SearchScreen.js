import React, { useMemo } from 'react';
import queryString from 'query-string'
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse( location.search );

    const [ formValues, handleInputChange ] = useForm({
        search: q
    });

    const { search } = formValues;
    const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ]);

    const handleSubmit = (e) => {
        e.preventDefault();
        
        history.push( `?q=${ search }` )
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>

                    <form onSubmit={ handleSubmit }>
                        <input type="text"
                                autoComplete="off"
                                name="search" 
                                placeholder="Find your Hero"
                                className="form-control"
                                value={ search }
                                onChange={ handleInputChange }/>

                        <button type="submit"
                                className="btn btn-block btn-outline-primary m-1">
                            Search
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        ( q === '' ) 
                            &&
                            <div className="alert alert-info">
                                Search a hero
                            </div>
                    }

                    {
                        ( q !== '' && heroesFiltered.length === 0 ) 
                            &&
                            <div className="alert alert-danger">
                                There aren`t a heroes with { q }
                            </div>
                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={ hero.id }
                                    heroe={ hero }/>
                        ))
                    }

                </div>
            </div>
        </div>
    )
}
