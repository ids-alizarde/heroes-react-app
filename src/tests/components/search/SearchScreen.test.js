import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';



describe('Pruebas en <SearchScreen />', () => {
    
    test('Debe de mostrarse correctamente el componente <SearchScreen /> con los valores por defautl ', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search' ]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( '.alert-info' ).text().trim() ).toBe( 'Search a hero' );
    });

    test('Debe mostrar a Batman y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find( 'input' ).props().value ).toBe( 'batman' );
        expect( wrapper.find( 'HeroCard' ).exists() ).toBe( true );
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('Debe de mostrar un error si no existe el Hero', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=batman123123123' ]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find( '.alert-danger' ).exists() ).toBe( true );
        expect( wrapper.find( '.alert-danger' ).text().trim() ).toBe( 'There aren`t a heroes with batman123123123' );
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe llamar history.push()', () => {
        
        const historyMock = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <Route path="/search" component={ () => <SearchScreen history={ historyMock } /> }/>
            </MemoryRouter>
        );

        wrapper.find( 'input' ).simulate( 'change', {
            target: {
                name: 'search',
                value: 'batman'
            }
        });

        // wrapper.find( 'input' ).prop( 'onChange')( ({
        //     target: {
        //         name: 'search',
        //         value: 'batma'
        //     }
        // }));

        // wrapper.find( 'input' ).props().onChange(( { 
        //     target: {
        //         name: 'search',
        //         value: 'bat'
        //     }     
        // }));

        wrapper.find( 'form' ).props().onSubmit({ preventDefault(){} });
        
        expect( historyMock.push ).toHaveBeenCalled();
        expect( historyMock.push ).toHaveBeenCalledWith( '?q=batman' );
    });
    
})
