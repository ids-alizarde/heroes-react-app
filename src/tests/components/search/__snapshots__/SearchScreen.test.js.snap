// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar un error si no existe el Hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman123123123"
        />
        <button
          className="btn btn-block btn-outline-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There aren\`t a heroes with 
        batman123123123
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrarse correctamente el componente <SearchScreen /> con los valores por defautl  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-block btn-outline-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas....
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
